### PR TITLE
Board Support for Sparkfun WRL13990 (nrf52832)

### PR DESCRIPTION
**Pull request by rolandvs**
The WRL13990 is a minimum development board from Sparkfun.

It features an nRF52832 and a minimum of extra's like a button and a LED. The UART connector has the FTDI 6-pin format for easy access using the USB-RS232 cable of FTDI.

### CODE REVIEW

File: ports/nrf/boards/WRL13990/mpconfigboard.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#define MICROPY_HW_PWM1_NAME        "PWM1"
+#define MICROPY_HW_PWM2_NAME        "PWM2"
+
+#define HELP_TEXT_BOARD_LED         "1"
+
+// EXTRA FLASH (as used by Laird cs)
+// Connect Macronix MX25R4035F (4MBit) or MX25R8035F (8Mbit)
+#define MICROPY_HW_SPI1_NAME    	"SPI1"
+#define MICROPY_HW_SPI1_MISO		(20)
+#define MICROPY_HW_SPI1_CLK			(16)
```
Comment (aykevl) [inline ports/nrf/boards/WRL13990/mpconfigboard.h]: I see you have been using tabs here. MicroPython uses spaces for indentation.

File: ports/nrf/boards/WRL13990/mpconfigboard.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+#define MICROPY_PY_MACHINE_TIMER    (1)
+#define MICROPY_PY_MACHINE_RTCOUNTER (1)
+#define MICROPY_PY_MACHINE_I2C      (1)
+#define MICROPY_PY_MACHINE_ADC      (1)
+#define MICROPY_PY_MACHINE_TEMP     (1)
+#define MICROPY_PY_RANDOM_HW_RNG    (1)
+
+// options to check out
+//#define MICROPY_PY_BLE				(0)
+//#define MICROPY_HW_HAS_SDCARD		(0)
```
Comment (aykevl) [inline ports/nrf/boards/WRL13990/mpconfigboard.h]: This code (and other commented out code below) isn't used. Can you remove it, just for cleanliness?

Last time I tested SD card support using this flag doesn't work. `sdcard.py` probably works better.

**Review by aykevl — CHANGES_REQUESTED** [review]
Thank you for the contribution! I did a really quick review.

**Comment by aykevl** [comment]
Can you also add this board to the README?

File: ports/nrf/boards/WRL13990/mpconfigboard.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+//#define MICROPY_HW_UART1_RTS        (5)
+
+
+// SPI0 config
+#define MICROPY_HW_SPI0_NAME        "SPI0"
+#define MICROPY_HW_SPI0_SCK         (25)
+#define MICROPY_HW_SPI0_MOSI        (23)
+#define MICROPY_HW_SPI0_MISO        (24)
+
+// SPI1 config
```
Comment (glennrub) [inline ports/nrf/boards/WRL13990/mpconfigboard.h]: The defines for SPI1 would not have any effect as the machine/spi.c will only render SPI0 compile time. Does the Sparkfun breakout boards have a implicit board definition of where SPI1 is expected to be located? Like Arduino board layout expects SPI to be on pins located at the position of 11-13 and often CS on 10. If so, it could be an option to also update the modules/machine/spi.c to interpret a second compile time SPI. 

If there is no expectation on where SPI1 should be on the board i would probably have dropped this from the board definition as this would vary quite a lot between use cases of the board and would also require to implement the necessary hooks in the machine/spi.c to make them used. You can always define the pins used for mosi, miso, sck and cs from the machine.SPI module.

Comment (rolandvs) [reply ports/nrf/boards/WRL13990/mpconfigboard.h]: @glennrub - thanks for your remark. I will change it as indicated. There is no "assigned" SPI1 whatsoever so lets make it as simple as possible, but not simpler as a famous person once said :-).

File: ports/nrf/boards/WRL13990/pins.csv [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+P23,P23
+P24,P24
+P25,P25
+P26,P26
+P27,P27
+P28,P28
+P29,P29
+P30,P30
+P31,P31
+BTN,P6
```
Comment (glennrub) [inline ports/nrf/boards/WRL13990/pins.csv]: Could it be an option to replace the template definitions on top with the new functionality names and get them listed in ascending order? 

Also, if there is a need for fixing the position of SPI1 it would be good to align the names between SPI0 and SPI1, CLK and SCK.

Comment (rolandvs) [reply ports/nrf/boards/WRL13990/pins.csv]: I would like to keep the generic names P0..Pxx as well as the labeled pins. Maybe just move all the labeled pins up (first to see) and the rest of the list to follow. Having both generic and labeled names in the list is a pattern that is also used in the STM32 boards csv files.

So with the STM32 in mind I would leave the pin order as is.

Comment (glennrub) [reply ports/nrf/boards/WRL13990/pins.csv]: If i recall correctly the duplicate definition would not have any effect. The make-pins.py for nrf-port is a bit different compared to the stm32 port.

The expectation is that, 1. column will generate the name for Pin.board, and the 2. column for Pin.cpu. This means that a definition of 'BTN,P6' should render Pin.board.BTN and Pin.cpu.P6. This will keep both the generic name from the CPU and the unique name for the Board.

What happens for the duplicate is a bit unknown. Does it run on target, can you do the lookup properly?

File: ports/nrf/README.md [hunk 5]
```diff
@@ -106,6 +107,7 @@ wt51822_s4at         | s110                    | Peripheral             | Manual
 pca10040             | s132                    | Peripheral and Central | [Segger](#segger-targets)
 feather52            | s132                    | Peripheral and Central | Manual, SWDIO and SWCLK solder points on the bottom side of the board
 arduino_primo        | s132                    | Peripheral and Central | [PyOCD](#pyocdopenocd-targets)
+WRL13990             |                         | Peripheral and Central | [Segger](#segger-targets)
```
Comment (glennrub) [inline ports/nrf/README.md]: it would be great to keep the board name in lower case here to keep some kind of style. Also, i'm very confident that Bluetooth LE will run on this board, so you could point it to the s132 like the other nrf52832 targets as that will be the proper SoftDevice for this chip.

File: ports/nrf/boards/WRL13990/README.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
+# Nordic nRF5x micropython
```
Comment (glennrub) [inline ports/nrf/boards/WRL13990/README.md]: I think this page could be a good supplement for describing the board. For me it would made sense to skip this top level nRF5x title, keep the `WRL13990 Board Support` as title. The section below describes what chip is on the board.

File: ports/nrf/boards/WRL13990/README.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+A4       |P28|P28| Alternatively ADC0_IN4
+A5       |P29|P29| Alternatively ADC0_IN5
+A6       |P30|P30| Alternatively ADC0_IN6
+A7       |P31|P31| Alternatively ADC0_IN7
+
+
+
+## Example
+
+To get a quick overview of the GPIO pins, a module `show_pins().py` was added in the `nrf/examples` directory. All pins are listed (omitted a few here).
```
Comment (glennrub) [inline ports/nrf/boards/WRL13990/README.md]: I'm thinking that this additional _show_pins_ example should probably be a separate commit being added before this description is added. And if accepted, this section should could point to some example that exists in the past. 

Instead of `was added`it would be more in the terms of `use the script show_pins.py` located in the examples/ folder.

**Comment by glennrub** [comment]
@rolandvs , i am not sure if it makes sense to update the main.c and add a new example as part of board support. Would it be an idea to split the commit series into two PR's, first adding the board support itself?

The pure board support part of the PR, with a few more modifications (docs), starts looking OK to me with the latest updates on the `pins.csv` and `mpconfigboard.h`.

For the main.c update, i am wondering if the script in main.c,  for listing pins, is worth the flash size it will consume. is this something that all targets would need as inbuilt? I'm not opposing adding a helper script in examples/ if it makes sense, but my opinion is that main.c should be kept small. Even the LED script which is already present is a bit big to me, but kept in order to show where users could inject code for their own use cases when building from source. And, also show that the board is alive after flashing.

File: ports/nrf/boards/WRL13990/README.md (hunk 7 above)
Comment (rolandvs) [reply ports/nrf/boards/WRL13990/README.md]: I could remove it al together and just list the script in the readme and leave it to the reader to copy and paste it for his own benefit?

**Comment by rolandvs** [comment]
@glennrub - I REMOVED the addition to the `main.c` but in a way that this is really *lost*. Apart from history reasons, it is not a problem :-). Also removed the example `show_pins.py` and put it into the board `README.md` file. Changed the `nrf/README.md` so I think "it is ready" :-).

File: ports/nrf/boards/WRL13990/README.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+## Board Pins
+An overview of all available pins is in the table below. Most of the pins can be used for PWM, UART, SPI, I2C, etc. However some pins are special like the ADC and NFC provision. Please refer to the datasheet for additional information:
+
+- When using NFC pins take special care about configuration. See note in [datasheet](http://infocenter.nordicsemi.com/pdf/nRF52832_PS_v1.4.pdf), page 17..18. At reset the pins are set to NFC mode!
+- When UART hardware handshaking is required, select the GPIO's you like and set them accordingly in the `mpconfigboard.h`
+
+
+LABEL | BOARD | CPU | REMARK
+--- | --- | --- | ---
+         |P0 |P0 | Reserved 32kHz XTAL
```
Comment (glennrub) [inline ports/nrf/boards/WRL13990/README.md]: I'm sorry i am nitpicking here... 
I feel the board/cpu columns share a bit redundant information. Does it make sense to document the board (silkscreen label) and the CPU pin as two distinct values if they will always be the same? I'm thinking that it could make sense to maybe combine these two columns. What do you think?

**Comment by glennrub** [comment]
As target name now has been aligned to lowercase as requested, the folder board/WRL13990 should also be named in lowercase to align. This makes it possible for the Makefile to find the correct board folder for when issuing `make BOARD=wrl13990`.

**Comment by rolandvs** [comment]
Funny I did change the directory by renaming it to the lower case version. I see it in my local repo and use `make BOARD=wrl13990`. So what could be wrong?

File: ports/nrf/boards/WRL13990/mpconfigboard.h [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+#define MICROPY_HW_LED_PULLUP       (1) 
+
+#define MICROPY_HW_LED1             (7)     // LED1
+
+// UART config
+#define MICROPY_HW_UART1_HWFC       (0)
+#define MICROPY_HW_UART1_RX         (26)    // default 115200
+#define MICROPY_HW_UART1_TX         (27)
+//#define MICROPY_HW_UART1_CTS        (7)
+//#define MICROPY_HW_UART1_RTS        (5)
```
Comment (aykevl) [inline ports/nrf/boards/WRL13990/mpconfigboard.h]: With flow control disabled, these pins are unused so the definitions can be removed.

**Comment by aykevl** [comment]
> Funny I did change the directory by renaming it to the lower case version. I see it in my local repo and use `make BOARD=wrl13990`. So what could be wrong?

You probably didn't commit/push the changes, see `git status`.
To help the move, you can rename the directory back to uppercase and rename it using `git mv`. That will move the directory and add stage the change at the same time in git.

File: ports/nrf/README.md [hunk 10]
```diff
@@ -114,13 +116,13 @@ Install the necessary tools to flash and debug using Segger:
 
 [JLink Download](https://www.segger.com/downloads/jlink#)
 
-[nrfjprog linux-32bit Download](https://www.nordicsemi.com/eng/nordic/download_resource/52615/16/95882111/97746)
+[nrfjprog linux-32bit Download](https://www.nordicsemi.com/eng/nordic/download_resource/52621/24/48793639/97746)
```
Comment (aykevl) [inline ports/nrf/README.md]: These changes should really have been done in a separate PR, as they're unrelated to this board.

File: ports/nrf/boards/WRL13990/README.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+SPI0_SCK |P25|
+RXD      |P26| Receive serial data (baudrate 115200)
+TXD      |P27| Transmit serial data
+A4       |P28| Alternatively ADC0_IN4
+A5       |P29| Alternatively ADC0_IN5
+A6       |P30| Alternatively ADC0_IN6
+A7       |P31| Alternatively ADC0_IN7
+
+## Example
+To get a quick overview of the GPIO pins, a module `show_pins.py` is in the section below. It gives you an overview of all the defined pins and the way they are configured.
```
Comment (aykevl) [inline ports/nrf/boards/WRL13990/README.md]: While a script to show pins could be useful, I agree with @glennrub that it should be put in the examples/ folder (in a separate pull request). It certainly shouldn't be put in this README as it is not board-specific (except for this particular pin mapping which is already included in this README).

File: ports/nrf/boards/WRL13990/README.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
+# wrl13990 Board Support
+
+The Sparkfun `WRL13990` is a very basic, no thrills, `nRF52832-QFAA` based board. It can be used as a DIL socket and eases the development of a Bluetooth Low Energy or other ARM Cortex-M4 applications.
```
Comment (aykevl) [inline ports/nrf/boards/WRL13990/README.md]: Nitpick: this sentence isn't right, it has one "... a ..." too much.

File: ports/nrf/boards/WRL13990/README.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
+# wrl13990 Board Support
```
Comment (aykevl) [inline ports/nrf/boards/WRL13990/README.md]: Nitpick: the rest of the document uses uppercase WRL13990 so the title should certainly use that.
Also, the SparkFun website doesn't mention this name, which might be confusing. In other places it appears to be called WRL-13990 (with a dash), so using WRL13990 here may be confusing.

**Comment by aykevl** [comment]
Some other comments:

 * It would be nice to link to the board README from the port README. See [this commit](https://github.com/aykevl/micropython/commit/30ce6be612fb52496352e7e391084dd4952a1363) for how it can be done.
 * The tables in the board README are messed up ([see here](https://github.com/rolandvs/micropython/blob/SPARKFUN_NRF52832/ports/nrf/boards/WRL13990/README.md)). For a guide, see: https://help.github.com/articles/organizing-information-with-tables/

File: ports/nrf/README.md (hunk 10 above)
Comment (rolandvs) [reply ports/nrf/README.md]: I merely updated the current ones with the new ones, but as this is a manual process pointing to the pages with the overview of the current and previous tools is even better.

File: ports/nrf/boards/WRL13990/mpconfigboard.h (hunk 9 above)
Comment (rolandvs) [reply ports/nrf/boards/WRL13990/mpconfigboard.h]: I tend to disagree as it is an useful reminder when adding it after all, so I like to keep these definitions around.

**Comment by rolandvs** [comment]
@aykevl @glennrub Should I squash all the above clutter into a single one?

**Comment by glennrub** [comment]
@rolandvs I guess a squash could be good in the end. It could also make sense to keep the script addition (examples/show_pins.py) as a separate commit as it is a bit unrelated to the board support itself. 

When that's said, I still have a few questions about the script to list pins. Where is the original source of this? My thinking is that if there is already a script that does this already (which was adopted to nrf), it might be a good idea to look if we can make it common instead of introducing a port specific variant of it.

**Comment by rolandvs** [comment]
About the script, @dhylands made it a while ago. There is a new one that is a bit more "pretty" print, which you will find in the `examples/` directory of micropython. The change I made had to do that the `dir(pyb.Pin)` or `dir(machine.Pin)` now also outputs `__class__` and `__name__`  that will break the original script. What it should do is give an overview of all the pins and preferably the selected `af()` functions to get an overview. For now I could drop it from the port so that can be finished off and see if there is a need to add the script again in this form or more advanced...

File: ports/nrf/README.md [hunk 14]
```diff
@@ -106,6 +107,7 @@ wt51822_s4at         | s110                    | Peripheral             | Manual
 pca10040             | s132                    | Peripheral and Central | [Segger](#segger-targets)
 feather52            | s132                    | Peripheral and Central | Manual, SWDIO and SWCLK solder points on the bottom side of the board
 arduino_primo        | s132                    | Peripheral and Central | [PyOCD](#pyocdopenocd-targets)
+[wrl13990](ports/wrl13990)| s132                    | Peripheral and Central | [Segger](#segger-targets)
```
Comment (glennrub) [inline ports/nrf/README.md]: A minor syntactic sugar detail, but could you expand the "Target board (BOARD)" with a few spaces, and the also add some blank spaces on the other targets so that the PIPE characters get aligned?

**Comment by glennrub** [comment]
> For now I could drop it from the port so that can be finished off and see if there is a need to add the script again in this form or more advanced...

@rolandvs, Sound good to me. If you drop the `show_pins.py` and the reference to it in the `wrl13990/README.md` that will make me happy. We could also raise an Issue to discuss what to do about the `examples/` script. It might trigger some changes related to `machine.Pin` in the nrf-port as well, in order to align the other way around, more towards the already existing script instead of adopting it to work for nrf-port.

File: ports/nrf/boards/wrl13990/README.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+The Sparkfun `WRL-13990` is a very basic, no thrills, `nRF52832-QFAA` based board. It can be used as a DIL socket and eases the development of Bluetooth Low Energy or other ARM Cortex-M4 applications.
+
+## Board Pins
+An overview of all available pins is in the table below. Most of the pins can be used for PWM, UART, SPI, I2C, etc. However some pins are special like the ADC and NFC provision. Please refer to the datasheet for additional information.
+
+- When using NFC pins take special care about configuration. See note in [datasheet](http://infocenter.nordicsemi.com/pdf/nRF52832_PS_v1.4.pdf), page 17..18. At reset the pins are set to NFC mode!
+- When UART hardware handshaking is required, select the GPIO's you like and set them accordingly in the `mpconfigboard.h`.
+
+
+BOARD | CPU | REMARK
```
Comment (glennrub) [inline ports/nrf/boards/wrl13990/README.md]: A minor syntactic sugar detail on this table as well. Could you expand the columns so that the PIPE characters get aligned? It makes it a bit easier to read when using a text editor.

Comment (rolandvs) [reply ports/nrf/boards/wrl13990/README.md]: Ok that should do it :-)

**Comment by glennrub** [comment]
@rolandvs , I realized i made you remove the link from the `nrf/README.md` to the `boards/wrl13990/README.md`. That was not my intention. I believe @aykevl comment is still valid to keep such a link. Given that the link to the board documentation is back, my comment was about aligning the tabbing in the section so that it looks nice to read. And i guess the `show_pins.py` could also be dropped for now as discussed previously. 

Also, @aykevl comment about he updated links to Segger download is a bit unrelated in context of adding a new board and should ideally have been a separate PR. I guess we can squash the commits (70a12750d9de7ae82a5218227253e50234e6fcbe, 4426c0de00aeb791f3215c60cceddb60d6461ece) into a single commit which is not part of the pure board support part of the PR, or simply just remove these changes and create a new PR for these. Right now, to get this board support PR forward i'm leaning towards the latter, simply just drop these Segger link changes for now as well.

To summarize, my suggestions would be:
* Add back the link to the `boards/wrl13990`in the `nrf/README.md`
* Align the | characters in the Target Boards table so that the table looks nice in a text editor
* Drop the show_pins.py script 
* Drop the Segger link updates

Does this sound reasonable?

**Comment by rolandvs** [comment]
This PR is closed and replaced by #3991. All suggestions are processed:
- Sample code to show the pin configuration is removed
- The nrfjprog links were updated using a PR #3964
